File: blue-sdk-ios/BlueSDK/Example/BlueSDK/AlarmStorage.swift
// AlarmStorage.swift
// BlueSDK Example - 闹钟配置本地持久化
// 使用 UserDefaults 存储闹钟槽位配置，退出界面后保留

import Foundation

/// 闹钟本地存储管理器
final class AlarmStorage {

    static let shared = AlarmStorage()

    private let key = "alarm_slots_v1"
    private let defaults = UserDefaults.standard

    private init() {}

    /// 保存单个闹钟槽位
    func save(slot: AlarmSlot) {
        var all = loadAll()
        all[slot.index - 1] = slot
        persist(all)
    }

    /// 保存所有闹钟槽位
    func saveAll(_ slots: [AlarmSlot]) {
        persist(slots)
    }

    /// 加载所有闹钟槽位（7个）
    func loadAll() -> [AlarmSlot] {
        guard let data = defaults.data(forKey: key),
              let stored = try? JSONDecoder().decode([StoredAlarm].self, from: data),
              stored.count == 7 else {
            return defaultSlots()
        }
        return stored.map { $0.toSlot() }
    }

    /// 清除指定槽位
    func clear(index: Int) {
        var all = loadAll()
        guard index >= 1, index <= 7 else { return }
        all[index - 1] = AlarmSlot(index: index, isEnabled: false, hour: 0, minute: 0, weekMask: 0x7F, isSet: false)
        persist(all)
    }

    /// 清除所有
    func clearAll() {
        persist(defaultSlots())
    }

    // MARK: - 私有

    private func defaultSlots() -> [AlarmSlot] {
        return (1...7).map {
            AlarmSlot(index: $0, isEnabled: false, hour: 0, minute: 0, weekMask: 0x7F, isSet: false)
        }
    }

    private func persist(_ slots: [AlarmSlot]) {
        let stored = slots.map { StoredAlarm(from: $0) }
        if let data = try? JSONEncoder().encode(stored) {
            defaults.set(data, forKey: key)
        }
    }
}

// MARK: - Codable 中间结构

private struct StoredAlarm: Codable {
    let index: Int
    let isEnabled: Bool
    let hour: Int
    let minute: Int
    let weekMask: Int
    let isSet: Bool

    init(from slot: AlarmSlot) {
        self.index = slot.index
        self.isEnabled = slot.isEnabled
        self.hour = slot.hour
        self.minute = slot.minute
        self.weekMask = slot.weekMask
        self.isSet = slot.isSet
    }

    func toSlot() -> AlarmSlot {
        return AlarmSlot(index: index, isEnabled: isEnabled, hour: hour, minute: minute, weekMask: weekMask, isSet: isSet)
    }
}

File: delivery/android/demo/app/src/main/kotlin/com/blue/demo/AlarmStorage.kt
// AlarmStorage.kt
// BlueSDK Demo - 闹钟配置本地持久化
// 使用 SharedPreferences 存储闹钟槽位配置，退出界面后保留

package com.blue.demo

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * 闹钟本地存储管理器
 * 使用 SharedPreferences + JSON 持久化 7 个闹钟槽位配置
 */
object AlarmStorage {

    private const val PREFS_NAME = "alarm_slots_prefs"
    private const val KEY_SLOTS = "alarm_slots_v1"

    private fun prefs(context: Context) =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    /** 保存单个闹钟槽位 */
    fun save(context: Context, slot: AlarmSlot) {
        val all = loadAll(context).toMutableList()
        if (slot.index in 1..7) {
            all[slot.index - 1] = slot
        }
        persist(context, all)
    }

    /** 保存所有闹钟槽位 */
    fun saveAll(context: Context, slots: List<AlarmSlot>) {
        persist(context, slots)
    }

    /** 加载所有闹钟槽位（7个） */
    fun loadAll(context: Context): List<AlarmSlot> {
        val json = prefs(context).getString(KEY_SLOTS, null) ?: return defaultSlots()
        return try {
            val arr = JSONArray(json)
            if (arr.length() != 7) return defaultSlots()
            (0 until 7).map { i ->
                val obj = arr.getJSONObject(i)
                AlarmSlot(
                    index = obj.getInt("index"),
                    isEnabled = obj.getBoolean("isEnabled"),
                    hour = obj.getInt("hour"),
                    minute = obj.getInt("minute"),
                    weekMask = obj.getInt("weekMask"),
                    isSet = obj.getBoolean("isSet")
                )
            }
        } catch (e: Exception) {
            defaultSlots()
        }
    }

    /** 清除指定槽位 */
    fun clear(context: Context, index: Int) {
        val all = loadAll(context).toMutableList()
        if (index in 1..7) {
            all[index - 1] = AlarmSlot(index, false, 0, 0, 0x7F, false)
        }
        persist(context, all)
    }

    /** 清除所有 */
    fun clearAll(context: Context) {
        persist(context, defaultSlots())
    }

    // MARK: - 私有

    private fun defaultSlots(): List<AlarmSlot> =
        (1..7).map { AlarmSlot(it, false, 0, 0, 0x7F, false) }

    private fun persist(context: Context, slots: List<AlarmSlot>) {
        val arr = JSONArray()
        for (slot in slots) {
            arr.put(JSONObject().apply {
                put("index", slot.index)
                put("isEnabled", slot.isEnabled)
                put("hour", slot.hour)
                put("minute", slot.minute)
                put("weekMask", slot.weekMask)
                put("isSet", slot.isSet)
            })
        }
        prefs(context).edit().putString(KEY_SLOTS, arr.toString()).apply()
    }
}
